### PR TITLE
Gradient Feature Selection (Ready to review)

### PR DESCRIPTION
**Pull request by xuehui1991**
Including two feature selector:
1. Gradient feature selector
2. GBDT selector

May refine docs in the future.

### CODE REVIEW

File: src/sdk/pynni/nni/feature_engineering/gradient_selector/fgtrain.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+#                   % (it, maxiter, int(epoch), t_left, v0))
+
+#             if writer is not None:
+#                 writer.add_scalar('ftrain', v0.clone().cpu().detach().numpy(),
+#                                   it)
+#                 writer.add_scalar('epoch', epoch, it)
+#                 writer.add_scalar('t', stop_conds['t'][-1], it)
+#                 if not np.isnan(stop_conds['df'][-1]):
+#                     writer.add_scalar('log10(dftrain)',
+#                                       np.log10(stop_conds['df'][-1]), it)
```
Comment (chicm-ms) [inline src/sdk/pynni/nni/feature_engineering/gradient_selector/fgtrain.py]: Remove this commented code?

Comment (xuehui1991) [reply src/sdk/pynni/nni/feature_engineering/gradient_selector/fgtrain.py]: done.

File: docs/en_US/FeatureEngineering/GBDTSelector.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+* **early_stopping_rounds** (int, require) - The early stopping setting in lightgbm. The detail you could reference [here](https://lightgbm.readthedocs.io/en/latest/Parameters.html).
+
+* **importance_type** (str, require) - could be 'split' or 'gain'. The 'split' means ' result contains numbers of times the feature is used in a model' and the 'gain' means 'result contains total gains of splits which use the feature'. The detail you could reference in [here](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.Booster.html#lightgbm.Booster.feature_importance).
+
+* **num_boost_round** (int, require) - number of boost round. The detail you could reference [here](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.train.html#lightgbm.train).
+
+**Requirement of `get_selected_features` FuncArgs**
+    
+    For now, the `get_selected_features` function has no parameters.
```
Comment (chicm-ms) [inline docs/en_US/FeatureEngineering/GBDTSelector.md]: I can see that there is a parameter `10` for `get_selected_features` in above example

File: docs/en_US/FeatureEngineering/GBDTSelector.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+...
+```
+
+And you could reference the examples in `/examples/trials/feature-selection/gbdt_selector/`, too.
+
+
+**Requirement of classArgs**
+
+    For now, classArgs has no parameters.
```
Comment (chicm-ms) [inline docs/en_US/FeatureEngineering/GBDTSelector.md]: why there is a section `Requirement of classArgs` ?

File: docs/en_US/FeatureEngineering/GradientFeatureSelector.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+# will return the index with important feature here.
+print(fgs.get_selected_features())
+
+...
+```
+
+And you could reference the examples in `/examples/trials/feature-selection/gradient_feature_selector/`, too.
+
+
+**Requirement of classArgs**
```
Comment (chicm-ms) [inline docs/en_US/FeatureEngineering/GradientFeatureSelector.md]: Term `classArgs` is not explained above,  maybe rename to `Parameters of class FeatureGradientSelector constructor` or something else?

File: src/sdk/pynni/nni/feature_engineering/gbdt_selector/gbdt_selector.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
+# NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+# DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT
+# OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+# ==================================================================================================
+
+"""
+gbdt_selector.py including:
+    class GBDTSelector
```
Comment (chicm-ms) [inline src/sdk/pynni/nni/feature_engineering/gbdt_selector/gbdt_selector.py]: I did not see this kind of information in other nni code,  maybe remove it?

File: docs/en_US/FeatureEngineering/GBDTSelector.md (hunk 2 above)
Comment (xuehui1991) [reply docs/en_US/FeatureEngineering/GBDTSelector.md]: You are right. I forget to update the document.

File: docs/en_US/FeatureEngineering/GBDTSelector.md (hunk 3 above)
Comment (xuehui1991) [reply docs/en_US/FeatureEngineering/GBDTSelector.md]: I could delete it if you don't like it.

File: docs/en_US/FeatureEngineering/GradientFeatureSelector.md (hunk 4 above)
Comment (xuehui1991) [reply docs/en_US/FeatureEngineering/GradientFeatureSelector.md]: Good suggestion.

File: src/sdk/pynni/nni/feature_engineering/gbdt_selector/gbdt_selector.py (hunk 5 above)
Comment (xuehui1991) [reply src/sdk/pynni/nni/feature_engineering/gbdt_selector/gbdt_selector.py]: OK.

File: docs/en_US/FeatureEngineering/GBDTSelector.md (hunk 3 above)
Comment (chicm-ms) [reply docs/en_US/FeatureEngineering/GBDTSelector.md]: Since there is no useful information in the section and term `classArgs` is not explained.  Maybe it is better to remove it.

File: src/sdk/pynni/nni/feature_engineering/gbdt_selector/gbdt_selector.py (hunk 5 above)
Comment (chicm-ms) [reply src/sdk/pynni/nni/feature_engineering/gbdt_selector/gbdt_selector.py]: OK, i did not notice that, and since the information is obvious, I think we do not need it.

Comment (xuehui1991) [reply src/sdk/pynni/nni/feature_engineering/gbdt_selector/gbdt_selector.py]: OK

File: examples/trials/feature-selection/gbdt_selector/gbdt_selector_test.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
+# Copyright (c) Microsoft Corporation
```
Comment (QuanluZhang) [inline examples/trials/feature-selection/gbdt_selector/gbdt_selector_test.py]: how about moving these examples to `examples/feature-selection/gbdt....`?

File: src/sdk/pynni/nni/feature_engineering/gradient_selector/constants.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+# DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT
+# OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+# ==================================================================================================
+
+
+import numpy as np
+
+
+class StorageLevel:
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/feature_engineering/gradient_selector/constants.py]: what are these constants used for?

File: src/sdk/pynni/nni/feature_engineering/gradient_selector/requirements.txt [hunk 8]
```diff
@@ -0,0 +1,5 @@
+numpy==1.14.3
+scikit-learn==0.20.0
+scipy==1.1.0
+torch==1.1.0
+pytest
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/feature_engineering/gradient_selector/requirements.txt]: better to use `nnictl package install` to install dependencies

File: src/sdk/pynni/nni/feature_engineering/gradient_selector/constants.py (hunk 7 above)
Comment (QuanluZhang) [reply src/sdk/pynni/nni/feature_engineering/gradient_selector/constants.py]: never mind

**Comment by QuanluZhang** [comment]
better to add unit test in this pr or another pr

File: examples/trials/feature-selection/gbdt_selector/gbdt_selector_test.py (hunk 6 above)
Comment (xuehui1991) [reply examples/trials/feature-selection/gbdt_selector/gbdt_selector_test.py]: OK

File: src/sdk/pynni/nni/feature_engineering/gradient_selector/requirements.txt (hunk 8 above)
Comment (xuehui1991) [reply src/sdk/pynni/nni/feature_engineering/gradient_selector/requirements.txt]: Does any compressor install their dependency by `nnictl package install`?  I could take them as example.

Comment (QuanluZhang) [reply src/sdk/pynni/nni/feature_engineering/gradient_selector/requirements.txt]: they do not require more dependencies, if users want to compress a pytorch model, we assume they have installed pytorch. It is better to use `nnictl package install` for convenience, but you can keep it for now.

Comment (xuehui1991) [reply src/sdk/pynni/nni/feature_engineering/gradient_selector/requirements.txt]: I want to keep it.

File: docs/en_US/FeatureEngineering/Overview.md [hunk 9]
```diff
@@ -0,0 +1,3 @@
+# FeatureEngineering
+
+We are glad to announce the alpha release for Feature Engineering toolkit on top of NNI, it's still in the experiment phase which might evolve based on usage feedback. We'd like to invite you to use, feedback and even contribute.
```
Comment (QuanluZhang) [inline docs/en_US/FeatureEngineering/Overview.md]: better to give more description, the design, supported algorithms, etc.
